kernel_test.go

```diff
@@ -0,0 +1,87 @@
+package cbpfc
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net"
+	"os"
+	"testing"
+	"time"
+	"unsafe"
+
+	"golang.org/x/net/bpf"
+	"golang.org/x/sys/unix"
+)
+
+// kernelBackend is a backend that runs cBPF in the kernel
+func kernelBackend(tb testing.TB, insns []bpf.Instruction, in []byte) XDPAction {
+	filter, err := bpf.Assemble(insns)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	// Use a unix socket to test the filter
+	// This doesn't risk interfering with any other network traffic, doesn't require / add special
+	// headers (as would be the case if we used UDP for example) that the XDP tests don't deal with,
+	// and doesn't require any special permissions
+
+	// Temporary file for unix socket
+	temp, err := ioutil.TempFile("", "cbpfc_test")
+	if err != nil {
+		tb.Fatal(err)
+	}
+	path := temp.Name()
+	if err := os.Remove(path); err != nil {
+		tb.Fatal(err)
+	}
+
+	read, err := net.ListenUnixgram("unixgram", &net.UnixAddr{Name: path, Net: "unixgram"})
+	if err != nil {
+		tb.Fatal(err)
+	}
+	readConn, err := read.SyscallConn()
+	if err != nil {
+		tb.Fatal(err)
+	}
+	err = readConn.Control(func(fd uintptr) {
+		err := unix.SetsockoptSockFprog(int(fd), unix.SOL_SOCKET, unix.SO_ATTACH_FILTER, &unix.SockFprog{
+			Len:    uint16(len(filter)),
+			Filter: (*unix.SockFilter)(unsafe.Pointer(&filter[0])),
+		})
+		if err != nil {
+			tb.Fatal(err)
+		}
+	})
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	write, err := net.Dial("unixgram", path)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	if _, err := write.Write(in); err != nil {
+		tb.Fatal(err)
+	}
+
+	read.SetDeadline(time.Now().Add(50 * time.Millisecond))
+
+	var out [1500]byte
+	n, err := read.Read(out[:])
+	if err != nil {
+		// A timeout indicates that the filter did not match
+		if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+			return XDPPass
+		}
+
+		tb.Fatal(err)
+	}
+
+	// Bytes we did capture should match input
+	if !bytes.Equal(in[:n], out[:n]) {
```

Comment:
If you re-slice `in` you won't detect truncated output?
```suggestion
	if !bytes.Equal(in, out[:n]) {
```

Reply:
SocketFilters truncate matching packets to the SocketFilter's return value, I added a better comment :)

>  // Received packet is truncated to the SocketFilter's return value

Reply:
So we want the filter to truncate the packet? Why doesn't the socket filter return the packet length, or some arbitrary length? I still don't get it TBH.

Reply:
For SocketFilters, the return value is interpreted by the kernel as the number of bytes of the packet to pass through. So if the filter returns:

* `0`: `0` bytes of the packet are passed though, it is effectively "dropped" by the kernel
* `13`: The packet is truncated to `13` bytes by the kernel

Typically, most filters either return `0` for "no-match" (ie drop), or `1500` for "match" (ie pass) without truncating the packet. For example, this is where `tcpdump`'s `snaplen` parameter comes into play, it sets the filter's return value for "match" (and so how much of each matching packet is captured).

The actual return value of the SocketFilter isn't exposed, so we can only check if we got a packet or not, even if it's been truncated to one byte - that just means the filter returned `1`.

Reply:
Ok, that part makes sense to me. However, since we control all parameters of the cBPF filter, shouldn't this always return the full packet? Where is the `snaplen` equivalent in cbpfc?

Looking at it from a different angle: if truncation is the only thing that is possible, and we don't control where stuff is truncated, why compare the contents in the first place? Anything above n == 0 means we got a packet, and that's all that matters.

Reply:
> Ok, that part makes sense to me. However, since we control all parameters of the cBPF filter, shouldn't this always return the full packet? Where is the snaplen equivalent in cbpfc?

The eBPF or C code generated by `cbpfc` provides the same return value the cBPF would have generated. It's up to users to interpret it as they see fit.
For the sake of our tests, we treat anything non-zero as a match. And while we could ensure our tests always return either 0 or 1500, user programs might not - it makes sense to be able to use arbitrary return values (even thought it's hard to check them for correctness right now).

> Looking at it from a different angle: if truncation is the only thing that is possible, and we don't control where stuff is truncated, why compare the contents in the first place? Anything above n == 0 means we got a packet, and that's all that matters.

It's just a sanity check to make sure everything is working as expected. I added more to the comment:

```
// Sanity check we received the right packet
// Received packet is truncated to the SocketFilter's return value
```

I can just remove it if you like.